cuML/DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,114 @@
+# cuML developer guide
+This document summarizes rules and best practices for contributions to C++ component cuML of RAPIDS/cuml.
+
+# General
+
+# Thread safety
+cuML should be thread safe and its functions can be called from multiple host threads if they use different handles.
+
+The implementation of cuML should be single threaded.
+
+# Coding style
+
+# Error handling
+All calls to CUDA APIs should be done via the provided helper macros `CUDA_CHECK`, `CUBLAS_CHECK` and `CUSOLVER_CHECK`. Those macros take care of checking the return values of the used API calls and generate an exception in case the command was no successful. In case an exception needs to be avoided, e.g. when implementing a destructor, `CUDA_CHECK_NO_THROW`, `CUBLAS_CHECK_NO_THROW ` and `CUSOLVER_CHECK_NO_THROW ` (currently not available, see https://github.com/rapidsai/cuml/issues/229) should be used. Those macros will only log the error but do not throw an exception.
+
+# Logging
+Add once https://github.com/rapidsai/cuml/issues/100 is addressed.
+
+# Device and Host memory allocations
+To enable `libcuml.so` users to control how memory for temporary data is allocated device memory should only be allocated via the allocator provided:
+```cpp
+template<typename T>
+void foo(const ML::cumlHandle_impl& h, cudaStream_t stream, ... )
+{
+    T* temp_h = h.getDeviceAllocator()->allocate(n*sizeof(T), stream);
+    ...
+    h.getDeviceAllocator()->deallocate(temp_h, n*sizeof(T), stream);
+}
+```
+the same rule applies to larger amounts of host heap memory:
+```cpp
+template<typename T>
+void foo(const ML::cumlHandle_impl& h, cudaStream_t stream, ... )
+{
+    T* temp_h = h.getHostAllocator()->allocate(n*sizeof(T), stream);
+    ...
+    h.getHostAllocator()->deallocate(temp_h, n*sizeof(T), stream);
+}
+```
+Small host memory heap allocations, e.g. as internally done by STL containers, are fine, e.g. an `std::vector` managing only a handful of integers.
+Both the Host and the Device Allocators might allow asynchronous stream ordered allocation and deallocation. This can provide significant performance benefits so a stream always needs to be specified when allocating or deallocating (see [Asynchronous operations and stream ordering](# Asynchronous operations and stream ordering)).
+There are two simple container classes compatible with the allocator interface `ML::device_buffer` available in `cuML/src/common/device_buffer.hpp` and `ML::host_buffer` available in `cuML/src/common/host_buffer.hpp`. These allow to follow the [RAII idiom](https://en.wikipedia.org/wiki/Resource_acquisition_is_initialization) to avoid resources leaks and enable exception safe code. These containers also allow asynchronous allocation and deallocation using the `resize` and `release` member functions:
+```cpp
+template<typename T>
+void foo(const ML::cumlHandle_impl& h, ..., cudaStream_t stream )
```

Comment:
I especially now like the idea about writing our algos as: `void cumlAlgo(const cumlHandle_impl& h, ...)`. This avoids a ton of `getImpl` calls in the downstream which would have been necessary, otherwise!

Reply:
Yup. Fixed it.